Fix set texture for the bone

Icosider opened the pull request:
I noticed that if you use the `getTextureForBone` method to set the texture of a particular bone, then the texture can be superimposed not only on the selected bone, but also on others(example: head, arm, etc.)

With fix:
![with-fix](https://user-images.githubusercontent.com/9529993/192541767-bb45395e-46e0-4e2b-9ec3-14d2a6d2241c.png)

Without fix:
![without-fix](https://user-images.githubusercontent.com/9529993/192541776-02b00b04-e36d-4b68-9d3e-dc8407475b75.png)

AzureDoom commented:
Gonna have @DerToaster98 check this out when they can as the extended entity is their brain child.

Icosider commented:
> Gonna have @DerToaster98 check this out when they can as the extended entity is their brain child.

I have attached screenshots as well.

DerToaster98 commented:
That's not how it is supposed to work. Your "fix" forces the entire thing to always use translucent, use getRenderTypeForBone() instead, you need to lookup the exact name.

It is not meant to "overlay" textures. It is meant to force the use of a certain texture for a bone. The rendertype to use there gets set via getRenderTypeForBone()

I'm on construction site currently, will be back on friday and will take a closer look then. But currently i don't think this is a bug

Icosider commented:
> That's not how it is supposed to work. Your "fix" forces the entire thing to always use translucent, use getRenderTypeForBone() instead, you need to lookup the exact name.
> 
> It is not meant to "overlay" textures. It is meant to force the use of a certain texture for a bone. The rendertype to use there gets set via getRenderTypeForBone()
> 
> I'm on construction site currently, will be back on friday and will take a closer look then. But currently i don't think this is a bug

Yes, I didn't mean overlaying the texture, but setting the texture for the bone. You can look at a screenshot of what happens if you set a different bone texture. I do not know what caused the problem of replacing the texture not for one particular bone, but for some others, but it seems this fix solves it.

Icosider commented:
Explanation of the screenshot:
I set the texture for a specific armor bone (in my case, this is the tail armor bone), but for some reason, the arms became transparent, although I indicated in the check the specific bone for which the texture should be set(tail armor bone)

DerToaster98 commented:
The problem here is your renderer. After it rendered the bone with a overridden texture it will use the rendertype that the rest of the renderer uses. Since you **don't** provide us *anything* we think that the issue here is your renderer. Your "fix" (like stated before) makes it **impossible** to use a different rendertype than translucent.

Icosider commented:
Why did you close the PR? I can provide code as well as models and textures to test this :/

DerToaster98 commented:
I closed it cause it doesn't "fix" anything. The system works as it should. YOur case here is more likely caused by your renderer not using the translucent rendertype. What you do there breaks rendering since it exchanges the buffer when it does not expect it.

If you want to continue it provide a link to your repository (renderer  class, model class, model json files and textures). next time provide that at the start please

DerToaster98 commented:
Reopening until test files are provided

Icosider commented:
[Link to repo](https://github.com/Icosider/EposRaces)

DerToaster98 commented:
Did you just straight up copy-paste what i made? Not cool, really not cool. At least credit it in the file itself or ask.

DerToaster98 commented:
Also i looked: You never fill out the method that sets the rendertype for your model. That is your exact problem, do that and it should work. I'm not planning to spend more time with someone who seemingly gives zero fucks about crediting other's work properly

Icosider commented:
> Did you just straight up copy-paste what i made? Not cool, really not cool. At least credit it in the file itself or ask.

The copy paste was made to fix the model specifically for the player, this code will not remain in the public, so this is not an actual "theft" of your code. You needn't worry about it...

Icosider commented:
I added RenderType#entityTranslument in the redefined getRenderType method:
![image](https://user-images.githubusercontent.com/9529993/192851788-9667c3b1-20aa-4844-9652-681f43d5c282.png)

DerToaster98 commented:
> > Did you just straight up copy-paste what i made? Not cool, really not cool. At least credit it in the file itself or ask.
> 
> The copy paste was made to fix the model specifically for the player, this code will not remain in the public, so this is not an actual "theft" of your code. You needn't worry about it...

Even if it does not remain: At least write a comment at the top of the class that indicates where you got it from.

DerToaster98 commented:
> I added RenderType#entityTranslument in the redefined getRenderType method: ![image](https://user-images.githubusercontent.com/9529993/192851788-9667c3b1-20aa-4844-9652-681f43d5c282.png)

looks like the correct method. Can't find that line in your repository though.

Icosider commented:
Try it now

DerToaster98 commented:
> Try it now

Ye, you committed it now. That should do it. If it doesn't i'll take a look myself when i find the time, i won't use this PR cause it breaks things.

Icosider commented:
At the moment, this solution does not work, unfortunately :C

Icosider commented:
This PR may need to be closed later if you find a solution to the problem and send it to the original repository yourself.

DerToaster98 commented:
> At the moment, this solution does not work, unfortunately :C

"Doesn't work" is pretty vague, what exactly doesn't work?

DerToaster98 commented:
> This PR may need to be closed later if you find a solution to the problem and send it to the original repository yourself.

Yes, though this isn't a pretty big problem honestly. I'll create a issue in the next few days when i took a proper look myself.

Just to get it clear: To reproduce it i could basically put a small cube inside a larger cube. Both cubes have their own bone. If the larger cube has it's own texture and that one has transparent parts it would not show the inner cube, is that correct? I'm not going to use your case cause it is too specific and i prefer to work with more general use scenarios.

DerToaster98 commented:
Anyway, will take a closer look tomorrow. Pretty tired right now, finally got back from construction site today

bernie-g commented:
If you copy or use Geckolib code in any way, you need to follow the MIT license. Specifically, I was not able to find a copy of the MIT license in your repository or any indication that you are copying GeckoLib code. We will not provide any more help until you address this issue. Thanks for understanding.

Icosider commented:
> If you copy or use Geckolib code in any way, you need to follow the MIT license. Specifically, I was not able to find a copy of the MIT license in your repository or any indication that you are copying GeckoLib code. We will not provide any more help until you address this issue. Thanks for understanding.

I added the license and added a link to the library to the readme.


> > At the moment, this solution does not work, unfortunately :C
> 
> "Doesn't work" is pretty vague, what exactly doesn't work?

To put it simply, nothing has changed since the RenderType was changed.

> Just to get it clear: To reproduce it i could basically put a small cube inside a larger cube. Both cubes have their own bone. If the larger cube has it's own texture and that one has transparent parts it would not show the inner cube, is that correct? I'm not going to use your case cause it is too specific and i prefer to work with more general use scenarios.

It's pretty weird to hear that :/ Specifically, to reproduce this problem, you can simply take a player model, put a cube on its body (as if it were armor), make a texture for the player model and the same for armor (textures should be of the same proportions) and then just set the texture for the "armor" bone. And maybe you will be able to get what I previously showed in the screenshots... I have specially attached the entire project with textures and a model so that you can immediately get this problem, rather than trying to find it.

DerToaster98 commented:
ok, i can't run that repository, never worked with loom and not planning to deal with it just for this.

Icosider commented:
As a temporary stub, this is quite enough. I am glad that you were able to identify the problem. I think it is worth closing the PR, since it does not carry practical benefits and it is worth opening an issue. If my repository is no longer needed, then I could delete it.

DerToaster98 commented:
I remembered something. Alpha channel textures don't properly work in mc. It is visible on items too, it is really picky about the order of the individual cubes to be rendered.

Icosider commented:
I see that you have made changes to the code, now everything is working as it should, or should I check it myself?

DerToaster98 commented:
You should check yourself.

I changed how some things are being handled. Keep the rendering order of cubes with a transparent texture in mind, those have to render "inner cube" => "outer cube".

Also from what i saw your handling of armor is really ... weird. It handles that automatically, no need to deal with additional outer bones that just have the armor texture applied, but that is just a hint.

AzureDoom commented:
This should now be better showcased on how to do with the new example entity added by Der, will be coming to the other versions outside of 1.16 Monday or so.
